### ISSUE TITLE
"no-unused-prop-types" inside async functions (include arrow)

### ISSUE THREAD

**Issue by isnifer** [yannickcr/eslint-plugin-react#1053]
Hi!
I think there is a bug with async functions.
**property**: "no-unused-prop-types"
**parser**: "babel-eslint@7.1.1"
**where**: "async functions"
**example**:
```js
{
  code: [
    'export class Example extends Component {',
    '  static propTypes = {',
    '    failed: PropTypes.func.isRequired,',
    '    loadUserProfile: PropTypes.func.isRequired,',
    '  }',
    '  handleLogin = async () => {',
    '    await this.props.failed();',
    '    await this.props.loadUserProfile();',
    '  };',
    '}'
  ].join('\n'),
  parser: 'babel-eslint'
},
```

**Comment by GGAlanSmithee** [comment]
Some additional info which might help:

this **_will_** error:
```javascript
onSubmit = async (e) => {
    this.setState({loggingIn: true, success: false, message: null})

    const {email, password} = this.state
    const {login, setAuthenticated} = this.props // not detected by eslint

    const {success, message} = await login(email, password)

    this.setState({loggingIn: false, success, message})
    setAuthenticated(success)

    if (success) {
        const {router, location: {state: {nextPathname} = {}}} = this.props

        setTimeout(() => {
            router.push(nextPathname || '/')
        }, 1000)
    }
}
```

this will **_not_** error:
```javascript
onSubmit = async (e) => {
    this.setState({loggingIn: true, success: false, message: null})

    const {email, password} = this.state
    // properly detected by eslint
    const {login, setAuthenticated, router, location: {state: {nextPathname} = {}}} = this.props

    const {success, message} = await login(email, password)
    this.setState({loggingIn: false, success, message})
    setAuthenticated(success)

    if (success) {
        setTimeout(() => {
            router.push(nextPathname || '/')
        }, 1000)
    }
}
```

The only difference between these snippets is how `this.props` is destructured. In the first example, `location` and `router` props are destructured separately, and in the second example all props are accessed together. This is probably not the whole solution, since @isnifer isn't destructuring at all. But mabe you can try and see if this works;

```javascript
{
  code: [
    'export class Example extends Component {',
    '  static propTypes = {',
    '    failed: PropTypes.func.isRequired,',
    '    loadUserProfile: PropTypes.func.isRequired,',
    '  }',
    '  handleLogin = async () => {',
    '    const {failed, loadUserProfile} = this.props;',
    '    await failed();',
    '    await loadUserProfile();',
    '  };',
    '}'
  ].join('\n'),
  parser: 'babel-eslint'
},
```

**Comment by TrevorSayre** [comment]
```
export default class Users extends Component {

  static propTypes = {
    ...
    onLoad: PropTypes.func.isRequired,
  };

  async componentWillMount() {
    const onLoad = this.props.onLoad;
    ...
  }

  ...

}
```

`''onLoad' PropType is defined but prop is never used (react/no-unused-prop-types)'`

**Comment by Wildhoney** [comment]
@TrevorSayre similar for me &mdash; using `eslint-plugin-react@7.1.0`.
